### ISSUE TITLE
Bokeh Server is incompatible with tornado=4.5

### ISSUE THREAD

**Issue by canavandl** [bokeh/bokeh#6152]
There's a currently undiagnosed issue where Bokeh Server applications are incompatible with the newest tornado v4.5 release.

The server appears to start normally, but doesn't receive any requests or open any sessions when queried.

To reproduce:
* Install Bokeh v0.12.5 via conda or pip
* Run a Bokeh Server app example via ``bokeh serve app.py``
(below is an example app)
* Try (and fail) to open an application session at http://localhost:5006/line_on_off in a browser or via cURL. You'll see that any request just hangs.

contents of app.py
```python
""" Example demonstrating turning lines on and off - with bokeh server

"""

import numpy as np

from bokeh.io import curdoc
from bokeh.layouts import row
from bokeh.palettes import Viridis3
from bokeh.plotting import figure
from bokeh.models import CheckboxGroup

p = figure()
props = dict(line_width=4, line_alpha=0.7)
x = np.linspace(0, 4 * np.pi, 100)
l0 = p.line(x, np.sin(x), color=Viridis3[0], legend="Line 0", **props)
l1 = p.line(x, 4 * np.cos(x), color=Viridis3[1], legend="Line 1", **props)
l2 = p.line(x, np.tan(x), color=Viridis3[2], legend="Line 2", **props)

checkbox = CheckboxGroup(labels=["Line 0", "Line 1", "Line 2"], active=[0, 1, 2], width=100)

def update(attr, old, new):
    l0.visible = 0 in checkbox.active
    l1.visible = 1 in checkbox.active
    l2.visible = 2 in checkbox.active

checkbox.on_change('active', update)

layout = row(checkbox, p)
curdoc().add_root(layout)
```

Platform: MacOS and Ubuntu 14.04
Bokeh = 0.12.5
Tornado = 4.5

**Comment by bdarnell** [comment]
The culprit is the change to the handling of `yield None`: this used to raise `BadYieldError`, now it's equivalent to `yield gen.moment`. There are two places in Bokeh that catch BadYieldError; both of these need to be updated with explicit checks for None.

**Comment by bryevdv** [comment]
Working up a PR now.

**Comment by bryevdv** [comment]
@bdarnell THANKS was there a deprecation and if so, when did it start? i.e. will updated code work for tornado >= 4.3 ? Or will version checks be needed?

**Comment by bdarnell** [comment]
No, there was no advance notice of the change. I assumed that taking something that never worked (i.e. it always raised an exception) and making it work wouldn't break anybody; I wasn't considering the fact that someone might be catching this exception and relying on it at runtime. 

You won't need a version check; looking for None will work for all versions of Tornado. 

```
@gen.coroutine
def yield_for_all_futures(result):
    """ Converts result into a Future by collapsing any futures inside result.

    If result is a Future we yield until it's done, then if the value inside
    the Future is another Future we yield until it's done as well, and so on.
    """
    while True:
        if result is None:
            break
        try:
            future = gen.convert_yielded(result)
        except gen.BadYieldError:
            # result is not a yieldable thing, we are done
            break
        else:
            result = yield future
    raise gen.Return(result)
```

This function is dangerous, BTW: lots of things other than Futures are accepted by `convert_yielded` (and have been ever since it was introduced). In particular, an empty list (or empty dict) is accepted by `convert_yielded` and resolves into another empty list, resulting in an infinite loop. It's better to only peel off one layer of `Future` at a time, and only once per maybe-coroutine function call.

**Comment by bryevdv** [comment]
@bdarnell unfortunately this is not the issue or at least not all of it. With this diff:
```
diff --git a/bokeh/util/tornado.py b/bokeh/util/tornado.py
index 45834a28b..1a939a397 100644
--- a/bokeh/util/tornado.py
+++ b/bokeh/util/tornado.py
@@ -18,6 +18,9 @@ def yield_for_all_futures(result):
     while True:
         try:
             future = gen.convert_yielded(result)
+            if future is None:
+                import pdb; pdb.set_trace()
+                raise gen.BadYieldError
         except gen.BadYieldError:
             # result is not a yieldable thing, we are done
             break
@@ -58,6 +61,9 @@ class _AsyncPeriodic(object):
             result = self._func()
             try:
                 callback_future = gen.convert_yielded(result)
+                if callback_future is None:
+                    import pdb; pdb.set_trace()
+                    raise gen.BadYieldError
             except gen.BadYieldError:
                 # result is not a yieldable thing
                 return sleep_future
``` 
an app still does not connect or load, but the pdb breakpoints are never hit, so it's not even getting to these points in the code.

**Comment by canavandl** [comment]
I think it's because ``gen.convert_yielded(None)`` returns a Future. You may have to check if ``result`` is None or not

**Comment by bryevdv** [comment]
Sure enough, this diff works:
```
diff --git a/bokeh/util/tornado.py b/bokeh/util/tornado.py
index 45834a28b..4f305a14e 100644
--- a/bokeh/util/tornado.py
+++ b/bokeh/util/tornado.py
@@ -17,7 +17,11 @@ def yield_for_all_futures(result):
     """
     while True:
         try:
+            if result is None:
+                raise gen.BadYieldError
             future = gen.convert_yielded(result)
+            if future is None:
+                raise gen.BadYieldError
         except gen.BadYieldError:
             # result is not a yieldable thing, we are done
             break
@@ -57,7 +61,11 @@ class _AsyncPeriodic(object):
             sleep_future = self.sleep()
             result = self._func()
             try:
+                if result is None:
+                    raise gen.BadYieldError
                 callback_future = gen.convert_yielded(result)
+                if callback_future is None:
+                    raise gen.BadYieldError
             except gen.BadYieldError:
                 # result is not a yieldable thing
                 return sleep_future
```

**Comment by bdarnell** [comment]
The second check in each pair is unnecessary - `convert_yielded` will never return None.

**Comment by bryevdv** [comment]
@bdarnell first I really appreciate your input. I tested the following diff against 4.4.2 and 4.5 and it seems to work great, and if I understand your earlier comment, is also safer. Do you agree?
```
diff --git a/bokeh/util/tornado.py b/bokeh/util/tornado.py
index 45834a28b..03a366098 100644
--- a/bokeh/util/tornado.py
+++ b/bokeh/util/tornado.py
@@ -16,13 +16,11 @@ def yield_for_all_futures(result):
     the Future is another Future we yield until it's done as well, and so on.
     """
     while True:
-        try:
-            future = gen.convert_yielded(result)
-        except gen.BadYieldError:
-            # result is not a yieldable thing, we are done
+        # result is not a yieldable thing, we are done
+        if not isinstance(result, gen.Future):
             break
-        else:
-            result = yield future
+        future = gen.convert_yielded(result)
+        result = yield future
     raise gen.Return(result)

 class _AsyncPeriodic(object):
@@ -56,13 +54,10 @@ class _AsyncPeriodic(object):
             # the period.
             sleep_future = self.sleep()
             result = self._func()
-            try:
-                callback_future = gen.convert_yielded(result)
-            except gen.BadYieldError:
-                # result is not a yieldable thing
+            if not isinstance(result, gen.Future):
                 return sleep_future
-            else:
-                return gen.multi([sleep_future, callback_future])
+            callback_future = gen.convert_yielded(result)
+            return gen.multi([sleep_future, callback_future])
         def on_done(future):
             if not self._stopped:
                 self._loop.add_future(invoke(), on_done)
```

**Comment by bryevdv** [comment]
Also seems to work with py2.7 and tornado 4.3

**Comment by bdarnell** [comment]
Checking for `isinstance(result, gen.Future)` is too narrow - it won't work with python 3.5 `async def` native coroutines (among other things, but native coroutines are the most important). I think you could use `inspect.isawaitable` on python 3.5+ and `backports_abc.isawaitable` on older versions of python, although that wouldn't be 100% compatible with what you were doing before. I'm not sure whether that would be a real problem or not. The solution that is most compatible with what you had before is the one from my comment above: https://github.com/bokeh/bokeh/issues/6152#issuecomment-294649755

**Comment by bryevdv** [comment]
@bdarnell OK thanks for the info, I have made a new PR that should just have the change you suggested.

FYI if you have the interest, any suggestions you have for the bokeh server in general would be very welcome!

**Comment by makmanalp** [comment]
Just hit this too - a little heads up in the examples repo could be very helpful!

**Comment by bryevdv** [comment]
Well, this has been fixed in master, the latest dev build does and the next release of Bokeh will work fine with Tornado 4.5 additionally the conda package in the defaults channel has had its dependencies updated so it will not install with 4.5 any longer. That still leaves the pip package, If someone knows a quick/surgical way to update the dependencies without cutting a new release, I'd be open to doing that. I'm just a little wary of adding info to the examples page that will quickly ne out of date.

**Comment by nickmachairas** [comment]
Hello everyone! First time commenting but I've been closely following development, I'm a big supporter of the bokeh package, fantastic job done.

I am having problems as well with updating multiple plots with a callback function arranged with the layout function. I was getting DeserializationErrors in version 0.12.2, bokeh serve was hanging in 0.12.5. However, the plots are updating with the callback in dev version 0.12.6 but I had to nest each plot separately in column functions. It works but I see multiple RuntimeErrors in the console.

I can help with testing if needed. Thanks

**Comment by mjmare** [comment]
A notice in the user guide or somewhere else would have saved me a few hours of head scratching...

**Comment by makmanalp** [comment]
I also hit this *again* today, somehow didn't recall that I hit it earlier, and spent an hour testing AWS firewall, instance firewall, nginx proxy config, bokeh host config, switching out with another bokeh app, etc. I do agree with @mjmare as the bug is unusually insidious (there is no indication that anything is going wrong at all, plus also seems to work on my machine but not the server!) and not easily googlable. Checking the github issues should have been my first instinct, probably.

**Comment by tdsmith** [comment]
Hello from PyCon; I ran into this in a workshop this morning :) Pushing a .post0 release to PyPI with a constrained install_requires for tornado would be super helpful.

**Comment by bryevdv** [comment]
> Pushing a .post0 release to PyPI with a constrained install_requires for tornado would be super helpful.

Not really possible, I am afraid. Bokeh is a dual runtime library and has to coordinate separate JavaScript resources located separately on CDN. There is code to compute the correct CDN URL for these resources from the current version, as long as the version conforms to understood conventions. `.post0` does not follow the conventions, and things would almost certainly fall over with a version like that. Changing any of that  machinerywould be an unacceptable risk. There will be a new version of Bokeh out in approximately two weeks that works with Tornado 4.5. 

As I said if someone knows of a way to update the dependencies without changing the version *at all*, I would be amendable to trying that (with some assistance)

**Comment by mattpap** [comment]
> update the dependencies without changing the version at all,

I believe there should be no such way. If you change dependencies, you effectively change the package and this should be somehow reflected in package's identification. At least build number should differ. I'm a bit surprised that conda gives me this:
```
conda search bokeh | grep 0.12.5
                             0.12.5                   py27_0  defaults        
                             0.12.5                   py36_0  defaults        
                             0.12.5                   py35_0  defaults
```

**Comment by jni** [comment]
@bryevdv the conda-forge patch, I believe, didn't work. I was just affected by this bug in a conda env that I created a couple of days ago. It includes bokeh and distributed from conda-forge (no mention of tornado in the yml). tornado 4.5* was installed in the env.

**Comment by yonatanp** [comment]
For anyone looking for a workaround - downgrading to tornado v4.3.3 worked for me.
It's recommended to run `pip freeze` first and keep track of your package set before you go down the rabbit hole.

**Comment by bryevdv** [comment]
At this point you can also upgrade to Bokeh 0.12.6, which resolved the incompatibility

**Comment by rafonseca** [comment]
Hi guys. I've just started working with bokeh server and ran into this problem or a similar one. I have something similar to that texas choropleth example. With Tornado 4.5 and Bokeh 0.12.6, the connection closes ("WebSocket connection closed: code=None, reason=None") when I zoom or pan it. 
I downgraded to Tornado 4.3 and it does work now.

**Comment by parekhnish** [comment]
Hi! I am working on Bokeh 0.12.7 on Python 3.6.2, inside the conda environment. I was running into the same issue with Tornado 4.5.2. My workaround was downgrading to Tornado 4.4.3, which is from the conda-forge channel (and NOT the default channel). Hope this helps.

**Comment by Hrobinsonphilips** [comment]
I'm also having the same problem as above - using Bokeh 0.12.7, Python 3.6.2 and Tornado 4.5.2

**Comment by bryevdv** [comment]
If the app is showing anything thing other than a completely blank screen, with a stuck loading bar, or if the Websocket connects *at all* then this is *not* the same issue as the original issue. Please open s separate, *new* issue with complete information and code sample.

**Comment by mgmoulton** [comment]
I have the same problem on the Cloud9 linux machine. Using Bokeh 0,12,10 ... downgraded Tornado from 4.5.2 to 4.3 .... still have the problem ,,,

**Comment by bryevdv** [comment]
@mgmoulton then you have a different problem. I use Bokeh with tornado 4.5.2 daily. The issue with Tornado 4.5 was resolved last year. Please open a *new* issue, with *complete details*
